test/provider/AppWindow.unittest.ts

```diff
@@ -0,0 +1,164 @@
+    });
+
+    test('A freshly-initialized window returns false', () => {
+        expect(testAppWindow.hasContextListener()).toBe(false);
+    });
+
+    test('Calling addContextListener once makes the query return true', () => {
+        testAppWindow.addContextListener();
+        expect(testAppWindow.hasContextListener()).toBe(true);
+    });
+
+    test('Calling addContextListener several times makes the query return true', () => {
+        for (let i = 0; i < 5; i++) {
+            testAppWindow.addContextListener();
+        }
+        expect(testAppWindow.hasContextListener()).toBe(true);
+    });
+
+    test('Calling removeContextListener on a fresh window has no effect', () => {
+        testAppWindow.removeContextListener();
+        expect(testAppWindow.hasContextListener()).toBe(false);
+    });
+
+    test('Calling addContextListener then removeContextListener makes the query return true then false', () => {
+        testAppWindow.addContextListener();
+        expect(testAppWindow.hasContextListener()).toBe(true);
+        testAppWindow.removeContextListener();
+        expect(testAppWindow.hasContextListener()).toBe(false);
+    });
+
+    test('The state of one TestAppWindow does not affect another', () => {
+        const secondTestAppWindow = new TestAppWindow(fakeIdentity, fakeAppInfo, mockChannel, Date.now(), 0);
+        secondTestAppWindow.addContextListener();
+
+        expect(testAppWindow.hasContextListener()).toBe(false);
+    });
+});
+
+describe('When querying if a window is ready to receive contexts', () => {
+    let testAppWindow: TestAppWindow;
+
+    beforeEach(() => {
+        // All tests in this section will use fake timers to allow us to control the Promise races precisely
+        jest.useFakeTimers();
+
+        testAppWindow = new TestAppWindow(fakeIdentity, fakeAppInfo, mockChannel, Date.now(), 0);
+    });
+
+    afterEach(() => {
+        jest.useRealTimers();
+    });
+
+    test('A window with a context listener already registered returns true immediately', async () => {
+        testAppWindow.addContextListener();
+
+        // Use a jest spy to track the timing of when the promise resolves without awaiting
+        const timingSpy = jest.fn();
+        testAppWindow.isReadyToReceiveContext().then(timingSpy);
+
+        // Do not advance time, but let any pending promises be actioned
+        await Promise.resolve();
+
+        // Promise should have resolved immediately, so spy should have been invoked
+        expect(timingSpy).toHaveBeenCalledWith(true);
+    });
+
+    describe('When the window does not have a listener registered', () => {
+        test('If the window was created longer than the timeout in the past, the promise resolves false immediately', async () => {
+            // @ts-ignore Updating the creation time to test old windows
+            testAppWindow._creationTime = Date.now() - 10000;
```

Comment:
Cleaner to fastforward faketime rather than rewind the window?

Reply:
Yea, probably. I'm a bit wary of fake time and the ordering of when we create the window vs when we stop time, but I'll give I a go and see how it works.

Reply:
Your suggestion started working properly once I swapped in the fake time stuff from Notes.

Reply:
Looks good, but commented out code still in place